compiler/testData/codegen/bytecodeText/nullCheckOptimization/deterministicNotNullChecks.kt

```diff
@@ -1,4 +1,7 @@
 // FILE: test/CallableDescriptor.java
+// IGNORE_BACKEND_FIR: JVM_IR
+// Here FIR adds implicit NOT_NULL cast for `origin`, resulting in an assertion being added,
+// which is the correct (yet mismatching) behavior, according to https://youtrack.jetbrains.com/issue/KT-35656
```

Comment:
According to the comment below about `JVM_IR` behavior, I guess this is also related to expected type checking. One thing I'm confused is, as mentioned in TODO below, "there should be an assertion," but it will be a breaking change?